### PR TITLE
promtool: read from stdin if no filenames are provided in check rules

### PR DESCRIPTION
**Pull request by fgouteroux**
Hello,

I want to add stdin support in promtool check rules, like this
```
$ cat /tmp/fgx.yaml | promtool check rules
Checking stdin
  SUCCESS: 1 rules found
```

We could do it with `promtool check rules /dev/stdin` but it's not supported by all platform like windows.

It remove the check `ExistingFiles` in command line, but this checks is performed later.

```
$ promtool check rules /tmp/nofile.yaml /tmp/fgx.yaml
Checking /tmp/nofile.yaml
  FAILED:
/tmp/nofile.yaml: open /tmp/nofile.yaml: no such file or directory

Checking /tmp/fgx.yaml
  SUCCESS: 1 rules found
```

Let me know if it could be merged.

### CODE REVIEW

**Comment by roidelapluie** [comment]
On the other side, - is a valid file name.

What about we threat "no filename" as stdin instead?

**Comment by fgouteroux** [comment]
> On the other side, - is a valid file name.
> 
> What about we threat "no filename" as stdin instead?

You mean a string with "no filename" or an empty string ?

File: cmd/promtool/main.go [hunk 1]
```diff
@@ -715,9 +715,22 @@ func CheckRules(ls lintConfig, files ...string) int {
 }
 
 func checkRules(filename string, lintSettings lintConfig) (int, []error) {
-	fmt.Println("Checking", filename)
+	var rgs *rulefmt.RuleGroups
+	var errs []error
+	if filename == "-" || filename == "" {
+		var buf bytes.Buffer
+		tee := io.TeeReader(os.Stdin, &buf)
```
Comment (dgl) [inline cmd/promtool/main.go:722]: I think this was copied from `CheckMetrics`, this shouldn't need TeeReader, just `io.ReadAll` is enough.

**Comment by dgl** [comment]
> You mean a string with "no filename" or an empty string ?

Rather than an empty string, it should behave like this for no argument, i.e. `promtool check rules ""` is different to `promtool check rules`.

That way you should be able to keep the `.ExistingFile()`, but drop `.Required()`.

**Comment by fgouteroux** [comment]
@dgl I'm not sure that declaring empty string is slice is a good way to avoid modifying the current logic of files in `checkRules` function.

**Comment by fgouteroux** [comment]
@dgl any chance to get this merge ??

**Comment by fgouteroux** [comment]
Hello, no news about this ?

File: cmd/promtool/main.go [hunk 2]
```diff
@@ -690,6 +690,11 @@ func CheckRules(ls lintConfig, files ...string) int {
 	failed := false
 	hasErrors := false
 
+	// add empty string to avoid matching filename
```
Comment (roidelapluie) [inline cmd/promtool/main.go:693]:
```suggestion
	// Add empty string to avoid matching filename.
```

File: cmd/promtool/main.go [hunk 3]
```diff
@@ -131,8 +131,8 @@ func main() {
 	checkRulesCmd := checkCmd.Command("rules", "Check if the rule files are valid or not.")
 	ruleFiles := checkRulesCmd.Arg(
 		"rule-files",
-		"The rule files to check.",
-	).Required().ExistingFiles()
+		"The rule files to check, default is read from standard input (STDIN).",
```
Comment (roidelapluie) [inline cmd/promtool/main.go:134]:
```suggestion
		"The rule files to check, default is read from standard input.",
```

I don't think it is stdin on all platforms (e.g. windows)

**Comment by roidelapluie** [comment]
Can you update the docs so the CI is happy?

**Comment by fgouteroux** [comment]
> Can you update the docs so the CI is happy?

yes

File: cmd/promtool/main.go [hunk 4]
```diff
@@ -690,6 +690,11 @@ func CheckRules(ls lintConfig, files ...string) int {
 	failed := false
 	hasErrors := false
 
+	// Add empty string to avoid matching filename.
+	if len(files) == 0 {
```
Comment (roidelapluie) [inline cmd/promtool/main.go:694]: I don't really like this.

File: cmd/promtool/main.go [hunk 5]
```diff
@@ -715,9 +720,23 @@ func CheckRules(ls lintConfig, files ...string) int {
 }
 
 func checkRules(filename string, lintSettings lintConfig) (int, []error) {
```
Comment (roidelapluie) [inline cmd/promtool/main.go:722]: Do you think we can take a slice if files here, and check if len == 0 for stdin?

**Review by roidelapluie — APPROVED** [review]

**Comment by roidelapluie** [comment]
Thanks!

**Comment by kfdm** [comment]
It would seem that this PR has broken checking from files.
Running `promtool check rules <filename>` returns nothing, while something like `promtool check rules < <filename>` shows the error in the file. I'm still trying to understand exactly what change has caused this and will try to file a more detailed issue later 🙇

**Pull request merged**